### PR TITLE
CON-168 Restrict an admin from creating super admins.

### PR DESCRIPTION
**Pull request by cop1fab**
#### Description
This Pull request restricts an admin from creating a super admin. 

#### Context of the problem to be solved. 
On our application, a super admin can be created by an admin, this PR ensures that an admin is restricted from creating a super admin. 

#### How should this be manually tested?
1. Clone the repo: `git clone https://github.com/andela/mrm_api.git`
2. Setup project according to Readme.md
3. checkout to branch `story/CON-168-stop-admin-to-create-super-admin`
4. run the following query to change the role of a user. Make sure the email is valid and already exists in our database and its role_id = 2 or 1

```
mutation{
  changeUserRole(email:"email@andela.com", roleId: 3){
    user{
      id
      roles{
        id
      }
    }
  }
}
```

#### Type of Change
- [x] New feature (non-breaking change which adds functionality

#### How has this been tested
- [x] Unit tests
- [x] Integration tests

#### Checklist
- [x] My changes generate no new warnings
- [x] My code follows the style guidelines of this project
- [x] I have linted my code prior to submission
- [x] Existing unit tests pass locally with my changes
- [x] Implementation works according to expectations

#### JIRA
[CON-168](https://andela-apprenticeship.atlassian.net/browse/CON-168)

### CODE REVIEW

**Comment by vic3king** [comment]
@cop1fab Please resolve conflicts here and try to update the status of this task.

File: alembic/versions/9d913959375e_migration_message.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.add_column('responses', sa.Column('state', sa.Enum('active', 'archived', 'deleted', name='statetype'), nullable=False))
+    # ### end Alembic commands ###
+
+
+def downgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.drop_column('responses', 'state')
+    # ### end Alembic commands ###
```
Comment (rajeman) [inline alembic/versions/9d913959375e_migration_message.py]: What is this alembic file for?

File: alembic/versions/f565c548fe22_migration_message.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.add_column('responses', sa.Column('state', sa.Enum('active', 'archived', 'deleted', name='statetype'), nullable=False))
+    # ### end Alembic commands ###
+
+
+def downgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.drop_column('responses', 'state')
+    # ### end Alembic commands ###
```
Comment (rajeman) [inline alembic/versions/f565c548fe22_migration_message.py]: and this

File: api/analytics/all_analytics_query.py [hunk 3]
```diff
@@ -56,7 +56,6 @@ def resolve_all_analytics(self, info, **kwargs):
         admin_details = get_user_from_db()
         query = Role.get_query(info)
         admin_role = query.filter_by(id=admin_details.roles[0].id).first()
-
```
Comment (rajeman) [inline api/analytics/all_analytics_query.py]: since you did not add/remove code from this file, this change is unnecessary

File: api/user/schema.py [hunk 4]
```diff
@@ -159,6 +159,14 @@ def mutate(self, info, email, **kwargs):
         return DeleteUser(user=exact_query_user)
 
 
+def restrictingAdmin(new_role):
```
Comment (vic3king) [inline api/user/schema.py]: I think this function might be better placed in a helpers folder.

Comment (mwinel) [inline api/user/schema.py]: @cop1fab please add a docstring to give an overview of how this function works. Ensure to highlight the parameters it takes in and what it returns.

File: alembic/versions/9d913959375e_migration_message.py (hunk 1 above)
Comment (cop1fab) [reply alembic/versions/9d913959375e_migration_message.py]: Ah, earlier we tried to run migrations with @vic3king that alembic file might have been created that time.

Comment (IssaIan) [reply alembic/versions/9d913959375e_migration_message.py]: You should not push this file @cop1fab

File: api/user/schema.py (hunk 4 above)
Comment (cop1fab) [reply api/user/schema.py]: done

File: api/user/schema.py [hunk 5]
```diff
@@ -212,11 +216,11 @@ def mutate(self, info, **kwargs):
         location_id = kwargs['location_id']
         query_user = User.get_query(info)
         user = query_user.filter(
-          UserModel.state == "active", UserModel.email == email).first()
+            UserModel.state == "active", UserModel.email == email).first()
         if not user:
             raise GraphQLError("User not found")
         new_location = LocationModel.query.filter_by(
-          id=location_id, state="active").first()
+            id=location_id, state="active").first()
```
Comment (mnswaleh) [inline api/user/schema.py]: pay attention to changes u introduce on a file. eg adding extra white spaces on a file is not necessary

File: tests/test_user/test_change_user_role.py [hunk 6]
```diff
@@ -61,3 +64,13 @@ def test_change_user_role_with_already_assigned_role(self):
             change_user_role_with_already_assigned_role_mutation,
             change_user_role_with_already_assigned_role_mutation_response
         )
+
+        def test_restricting_admin_to_create_super_admin(self):
+            """
+            Testing restriction of an admin to create a super admin
+            """
```
Comment (mnswaleh) [inline tests/test_user/test_change_user_role.py]: I don't understand the use of this block

File: tests/base.py [hunk 7]
```diff
@@ -374,6 +377,18 @@ def func_wrapper(self):
...
+        user.roles.append(super_admin_role)
+        user.save()
+        db_session().commit()
+        return func(self)
+    return func_wrapper
+
+
 def change_test_user_role_to_super_admin(func):
     def func_wrapper(self):
         user_role = Role(role='Super Admin')
```
Comment (mnswaleh) [inline tests/base.py]: I believe these two wrappers perform the same task you could use the former

File: api/user/schema.py [hunk 8]
```diff
@@ -175,24 +176,27 @@ def mutate(self, info, email, **kwargs):
         query_user = User.get_query(info)
         active_user = query_user.filter(UserModel.state == "active")
         exact_user = active_user.filter(UserModel.email == email).first()
+
         if not exact_user:
             raise GraphQLError("User not found")
 
         new_role = RoleModel.query.filter_by(id=kwargs['role_id']).first()
+
```
Comment (mwinel) [inline api/user/schema.py]: @cop1fab please refer to comment on line 59 by @rajeman and ensure all unnecessary changes are fixed.

File: api/user/schema.py [hunk 9]
```diff
@@ -175,24 +176,27 @@ def mutate(self, info, email, **kwargs):
...
         if not new_role:
             raise GraphQLError('invalid role id')
 
         current_user_role = exact_user.roles[0].role
         if new_role.role == current_user_role:
             raise GraphQLError('This role is already assigned to this user')
 
+        restricting_admin(new_role.role)
```
Comment (rajeman) [inline api/user/schema.py]: @cop1fab Please change the name of this function to `check_andmin_restiction` or something in form of action.

File: tests/base.py [hunk 10]
```diff
@@ -374,6 +377,18 @@ def func_wrapper(self):
     return func_wrapper
 
 
+def change_admin_user_role_to_super_admin(func):
```
Comment (rajeman) [inline tests/base.py]: Why do you need this function? There is a similar one in this file.

File: tests/test_user/test_change_user_role.py (hunk 6 above)
Comment (rajeman) [reply tests/test_user/test_change_user_role.py]: Kindly provide an answer to the above question

File: api/user/schema.py (hunk 9 above)
Comment (cop1fab) [reply api/user/schema.py]: done

File: tests/base.py (hunk 10 above)
Comment (cop1fab) [reply tests/base.py]: @rajeman this wrapper is helping me to find if the role of the user who is trying to change another user to super admin is himself a super_Admin. The other function creates the super Admin, and if I use it, the super admin will be duplicated hence the test database will complain.

**Comment by IssaIan** [comment]
@cop1fab kindly run flake8 before pushing. This will show you the lines with errors.

File: tests/base.py [hunk 11]
```diff
@@ -253,6 +255,7 @@ class CommonTestCases(BaseTestCase):
         - admin_token_assert_in
         - user_token_assert_equal
         - user_token_assert_in
+        - super_admin_token_assert_in
```
Comment (rajeman) [inline tests/base.py]: @cop1fab What is the essence of line `258`?

**Comment by rajeman** [comment]
@mwinel You can check this PR again to confirm your requests have been resolved.
